### PR TITLE
client: Use protobuf for SDN clients

### PR DESCRIPTION
**Pull request by smarterclayton**

### CODE REVIEW

**Comment by danwinship** [comment]
So does this work now? We had intentionally disabled protobuf support before because it didn't work with CRD types.

If it does work you probably need an update to `pkg/openshift-network-controller/` too; this patch will only fix sdn-node, not sdn-controller.

**Comment by smarterclayton** [comment]
You should only have disabled protobuf for crd resources.  Right now 1/2 of CPU time in SDN is json decoding of pods and nodes, which is roughly a tenth of a core per machine.

The client should properly negotiate json watches for CRDs in 1.16 but I need to verify there are no writes being done.

**Comment by smarterclayton** [comment]
The backport for 4.2 and 4.1 will use proto for core resources and json for SDN resources.

**Comment by smarterclayton** [comment]
SDN controller isn’t showing up in profiles yet, I’m mostly trying to get per node cpu back under control.

**Comment by smarterclayton** [comment]
If OVN has done the same thing we need to fix that as well (1.16 plus, should look like this, 1.16 minus should have proto for core resources)

File: pkg/openshift-sdn/proxy.go [hunk 1]
```diff
@@ -215,13 +216,25 @@ func (sdn *OpenShiftSDN) runProxy(waitChan chan<- bool) {
...
+		sdn.ProxyConfig.MetricsBindAddress = "*:9665"
+	}
 	if len(sdn.ProxyConfig.MetricsBindAddress) > 0 {
 		mux := http.NewServeMux()
 		mux.HandleFunc("/proxyMode", func(w http.ResponseWriter, r *http.Request) {
 			fmt.Fprintf(w, "%s", sdn.ProxyConfig.Mode)
 		})
 		mux.Handle("/metrics", prometheus.Handler())
+
+		mux.HandleFunc("/debug/pprof/", pprof.Index)
```
Comment (squeed) [inline pkg/openshift-sdn/proxy.go]: This port is open to the whole cluster - is there an information leak here?

Comment (danwinship) [reply pkg/openshift-sdn/proxy.go]: note the "DO NOT MERGE" on that commit :slightly_smiling_face:

Comment (smarterclayton) [reply pkg/openshift-sdn/proxy.go]: :) yeah I’m trying to look at what else might be broken.  Protobuf is the cheap easy win, but I’m seeing a lot of time in GC

**Comment by smarterclayton** [comment]
It looks like the biggest win is to not use the hybrid proxier if we don't use it - about 15-20% of CPU time is spent syncing the user space proxy rules.

It looks like most of the remaining time is in exec to iptables.

Removed the debugging commit, I think this is sufficient on its own for a small win.  PTAL

File: pkg/openshift-sdn/proxy.go (hunk 1 above)
Comment (squeed) [reply pkg/openshift-sdn/proxy.go]: It's not (wasn't) on the PR :-)

**Comment by squeed** [comment]
/approve
/lgtm

If you want to file a JIRA card, we can start looking in to ways to make the unidling proxy do nothing faster.

**Comment by smarterclayton** [comment]
/retest

I guess with unidling we have to sort through what the OVN strategy will be.

**Comment by danwinship** [comment]
> I guess with unidling we have to sort through what the OVN strategy will be.

OVN will do unidling natively, not with the hybrid proxy

**Pull request merged**